python/lsst/ts/eas/m1m3_thermal.py

```diff
@@ -0,0 +1,208 @@
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+import asyncio
+import logging
+import os
+
+from astropy.time import Time
+from lsst.ts import salobj
+
+
+class ControlLoopManager:
+    def __init__(self, domain: salobj.Domain, log: logging.Logger):
+        """
+        Initializes the ControlLoopManager.
+
+        Parameters
+        ----------
+        domain: salobj.Domain
+            The SALObj domain instance to use.
+
+        log: logging.Logger
+            A logger for log messages.
+        """
+        self.domain = domain
+        self.log = log
+        self.m1m3ts = salobj.Remote(domain, "MTM1M3TS")
+        self.ess = salobj.Remote(domain, "ESS", index=112)
+        self.task = None
+        self.stop_event = asyncio.Event()
+
+        os.environ["LSST_DDS_RESPONSIVENESS_TIMEOUT"] = "15s"
+
+    async def stop_control_loop(self) -> None:
+        """
+        Cancels all running control tasks and waits for them to terminate.
+        """
+        self.log.info("Stopping control loop...")
+        self.stop_event.set()
+        if self.task is not None:
+            self.task.cancel()
+            await asyncio.gather(self.task, return_exceptions=True)
+
+    async def run_control_loop(self, heaterdemand: list[int], fandemand: list[int]):
+        """Runs the control loop for the fans and the heaters.
+
+        Parameters
+        ----------
+        heaterdemand: list[int]
+            The heater power to be applied when the heater is turned on by
+            the MTM1M3TS.heaterFanDemand command. The values range from 0-255,
+            with 255 being 100%.
+        fandemand: list[int]
+            Fan RPM demand for the MTM1M3TS.headerFanDemand command, with
+            255 being 100%.
+        """
+        try:
+            # Wait for remotes to get set up...
+            await asyncio.sleep(5.0)
+
+            mixing = await self.m1m3ts.tel_mixingValve.next()
+            currentvalveposition = mixing.valvePosition
+            oldvalveposition = currentvalveposition
+
+            while True:
+                glycol = await self.m1m3ts.tel_glycolLoopTemperature.next()
+                mixing = await self.m1m3ts.tel_mixingValve.next()
+                fcu = await self.m1m3ts.tel_thermalData.next()
+                currenttemp = (
+                    glycol.insideCellTemperature1
+                    + glycol.insideCellTemperature2
+                    + glycol.insideCellTemperature3
+                ) / 3
+                currentvalveposition = mixing.valvePosition
+
+                fcu = await self.m1m3ts.tel_thermalData.next()
+                fanspeed = fcu.fanRPM
+                fcutemp = fcu.absoluteTemperature
+
+                airtemp = await self.ess.tel_temperature.next()
+                targettemp = airtemp.temperatureItem[0]
+
+                date = Time.now()
```

Comment:
you really don't need to log times, every message you send to the EFD has a timestamp associated with it.